### PR TITLE
Nav-v2

### PR DESCRIPTION
**Pull request by chrisnojima**
To ticket:
- [ ] chat slash / @ / # 
- [ ] signup doesn't auto switch you?
- [x] all settings flows
- [x] android
- [ ] old nav getpath calls
- [ ] try all flows
- [ ] wallet popup
- [ ] wallet flows
- [x] make a team fform chat info panel
- [ ] global errors
- [ ] global offline?
- [ ] upgraded modal header that matches one from stack

### CODE REVIEW

File: shared/actions/chat2/index.js [hunk 1]
```diff
@@ -1288,7 +1289,7 @@ function* messageSend(state, action) {
   // disable sending exploding messages if flag is false
   const ephemeralLifetime = Constants.getConversationExplodingMode(state, conversationIDKey)
   const ephemeralData = ephemeralLifetime !== 0 ? {ephemeralLifetime} : {}
-  const routeName = 'paymentsConfirm'
+  const routeName = 'chatPaymentsConfirm'
```
Comment (chrisnojima) [inline shared/actions/chat2/index.js]: lets make all routes uniquely named, using a prefix is likely best

File: shared/actions/platform-specific/index.desktop.js [hunk 2]
```diff
@@ -254,18 +254,6 @@ const setupEngineListeners = () => {
   })
 }
 
-function* loadStartupDetails() {
```
Comment (chrisnojima) [inline shared/actions/platform-specific/index.desktop.js]: this effectively happens in the reducer now

File: shared/actions/provision.js [hunk 3]
```diff
@@ -347,7 +347,7 @@ class ProvisioningManager {
   showCodePage = () =>
     RouteTreeGen.createNavigateAppend({
       parentPath: this._addingANewDevice ? devicesRoot : [Tabs.loginTab],
-      path: ['codePage'],
+      path: this._addingANewDevice ? ['deviceCodePage'] : ['codePage'],
```
Comment (chrisnojima) [inline shared/actions/provision.js]: there was a name clash here, both were called codePage

File: shared/actions/wallets.js [hunk 4]
```diff
@@ -205,7 +206,7 @@ function* requestPayment(state) {
   )
   const navAction = maybeNavigateAwayFromSendForm(state)
   yield Saga.sequentially([
-    ...(navAction ? [Saga.put(navAction)] : []),
```
Comment (chrisnojima) [inline shared/actions/wallets.js]: maybeNavigateAwayFromSendForm always returns an array that needs to be mapped now

File: shared/common-adapters/plain-input.desktop.js [hunk 5]
```diff
@@ -274,15 +274,15 @@ class PlainInput extends React.PureComponent<InternalProps> {
 }
 
 const StyledTextArea = styled.textarea(props => ({
-  '&::-webkit-inner-spin-button': {'-webkit-appearance': 'none', margin: 0},
+  '&::-webkit-inner-spin-button': {WebkitAppearance: 'none', margin: 0},
```
Comment (chrisnojima) [inline shared/common-adapters/plain-input.desktop.js]: fix for kebob case warning

File: shared/git/container.js [hunk 6]
```diff
@@ -52,17 +51,30 @@ const mapDispatchToProps = (dispatch: any, {navigateAppend, setRouteState, route
...
-    setRouteState({expandedSet: old.has(id) ? old.delete(id) : old.add(id)})
-  },
 })
 
-class GitReloadable extends React.PureComponent<{
-  ...{|_loadGit: () => void|},
-  ...React.ElementConfig<typeof Git>,
-}> {
+// keep track in the module
+let _expandedSet = I.Set()
```
Comment (chrisnojima) [inline shared/git/container.js]: keeping expanded state here and not in routeState

File: shared/router-v2/tab-bar/index.native.js [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+            ))}
+          </Kb.Box2>
+        </Kb.NativeSafeAreaView>
+      )
+    )
+  }
+}
+
+// TEMP to really know you're on this branch
+const backgroundColor = 'pink' // Styles.globalColors.darkBlue2,
```
Comment (chrisnojima) [inline shared/router-v2/tab-bar/index.native.js]: lets leave this in until we flip the flag

File: shared/settings/routes.desktop.js [hunk 8]
```diff
@@ -89,3 +94,59 @@ const routeTree = () => {
...
+  [Constants.notificationsTab]: {getScreen: () => require('./notifications/container').default},
+  changeEmail: {getScreen: () => require('./email/container').default},
+  changePassphrase: {getScreen: () => require('./passphrase/container').default},
+  dbNukeConfirm: {getScreen: () => require('./db-nuke-confirm/container').default},
+  deleteConfirm: {getScreen: () => require('./delete-confirm/container').default},
+  inviteSent: {getScreen: () => require('./invite-generated/container').default},
+  removeDevice: {getScreen: () => require('../devices/device-revoke/container').default},
+}
+
+class SettingsSubNav extends React.PureComponent<any> {
```
Comment (chrisnojima) [inline shared/settings/routes.desktop.js]: an example of a sub nav. in the old style we had childContainer. this replaces that (nicely i think)

**Comment by chrisnojima** [comment]
@keybase/react-hackers this is now ready. lets discuss

File: shared/chat/conversation/create-team-header/container.js [hunk 9]
```diff
@@ -14,12 +14,14 @@ const mapStateToProps = (state, {conversationIDKey}: OwnProps) => ({
...
-        {
-          props: {conversationIDKey},
-          selected: 'showNewTeamDialog',
-        },
-      ]})
+      RouteTreeGen.createNavigateAppend({
+        path: [
+          {
+            props: {conversationIDKey},
+            selected: 'chatShowNewTeamDialog',
```
Comment (MarcoPolo) [inline shared/chat/conversation/create-team-header/container.js]: This doesn't seem to work for me. from the info panel in chat and clicking turn this into a team

File: shared/router-v2/switcheroo.js [hunk 10]
```diff
@@ -28,4 +39,13 @@ class RouterSwitcheroo extends PureComponent<Props> {
   }
 }
 
-export default RouterSwitcheroo
+const mapDispatchToProps = dispatch => ({
+  persistRoute: path => dispatch(ConfigGen.createPersistRoute({path})),
+  updateNavigator: navigator => dispatch(ConfigGen.createSetNavigator({navigator})),
```
Comment (MarcoPolo) [inline shared/router-v2/switcheroo.js]: It would be nice if we didn't have this

File: shared/search/user-input/index.native.js [hunk 11]
```diff
@@ -203,7 +204,7 @@ class UserInput extends Component<Props, State> {
             }}
           >
             <TextInput
-              autoFocus={autoFocus}
+              autoFocus={flags.useNewRouter ? false : autoFocus}
```
Comment (MarcoPolo) [inline shared/search/user-input/index.native.js]: Why this change?

File: shared/people/container.js [hunk 12]
```diff
@@ -80,8 +103,21 @@ const mergeProps = (stateProps, dispatchProps) => {
...
 }
 
-export default connect<OwnProps, _, _, _, _>(
+const connected = connect<OwnProps, _, _, _, _>(
   mapStateToProps,
   mapDispatchToProps,
   mergeProps
 )(LoadOnMount)
+
+// $FlowIssue lets fix this
```
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply shared/people/container.js]: One idea, Add another param to namedConnect that would be navigationOptions

File: shared/wallets/routes.js [hunk 13]
```diff
@@ -114,3 +118,79 @@ const routeTree = () => {
...
+    }
+
+class WalletsSubNav extends React.PureComponent<any> {
+  render() {
+    const navigation = this.props.navigation
+    const index = navigation.state.index
+    const activeKey = navigation.state.routes[index].key
+    const descriptor = this.props.descriptors[activeKey]
+    const childNav = descriptor.navigation
+    const WalletsAndDetails = require('./wallets-and-details/container').default
```
Comment (MarcoPolo) [inline shared/wallets/routes.js]: Kinda weird that this is in the render. On purpose?

Comment (chrisnojima) [reply shared/wallets/routes.js]: I wanted to defer making this until it was needed.

File: shared/settings/routes.native.js [hunk 14]
```diff
@@ -72,3 +73,30 @@ const routeTree = () => {
...
+  [Constants.landingTab]: {getScreen: () => require('./about-container').default},
+  [Constants.notificationsTab]: {getScreen: () => require('./notifications/container').default},
+  [Constants.passphraseTab]: {getScreen: () => require('./passphrase/container').default},
+  [Constants.screenprotectorTab]: {getScreen: () => require('./screenprotector-container.native').default},
+  dbNukeConfirm: {getScreen: () => require('./db-nuke-confirm/container').default},
+  deleteConfirm: {getScreen: () => require('./delete-confirm/container').default},
+  inviteSent: {getScreen: () => require('./invite-generated/container').default},
+  privacyPolicy: {getScreen: () => require('./web-links.native').default},
+  removeDevice: {getScreen: () => require('../devices/device-revoke/container').default},
+  'tabs:settingsTab': {getScreen: () => require('./').default},
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/settings/routes.native.js]: This is short term cause it's the tabs constants. We can likely change it to something soon. Ultimately all the section routes should be renamed so they at least have a prefix (like settingsAboutPage or whatever and then these could be settingRoot or something)

File: shared/router-v2/switcheroo.js (hunk 10 above)
Comment (chrisnojima) [reply shared/router-v2/switcheroo.js]: Which aspect? We need a way to bridge redux and this. This is what they suggest right?

File: shared/router-v2/router.shared.js [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+          sa.push(StackActions.push({params: undefined, routeName: p}))
+        }
+      }
+
+      // switch the switch and do a reset of the stack
+      return sa.length
+        ? [switchStack, StackActions.reset({actions: sa, index: sa.length - 1})]
+        : [switchStack]
+    }
+    case RouteTreeGen.clearModals:
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/router-v2/router.shared.js]: Why mix these route tree actions and actual nav actions?

File: shared/router-v2/router.shared.js [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+      if (action.payload.path) {
+        const p = action.payload.path.last
+          ? action.payload.path.last()
+          : action.payload.path[action.payload.path.length - 1]
+
+        // a chat, we want people/inbox/chat
+        if (p === 'chatConversation') {
+          sa.push(StackActions.push({params: undefined, routeName: 'tabs:chatTab'}))
+          sa.push(StackActions.push({params: undefined, routeName: 'chatConversation'}))
+        } else if (p !== 'tabs:peopleTab') {
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/router-v2/router.shared.js]: This only happens when we load fresh right?

File: shared/router-v2/router.shared.js [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+      } else {
+        routeName = p.selected
+        params = p.props
+      }
+
+      if (!routeName) {
+        return
+      }
+      // don't allow pushing a dupe
+      const path = Constants.getVisiblePath(navigation.state)
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/router-v2/router.shared.js]: We can allow dupes within. We only disallow direct dupes

File: shared/router-v2/router.native.js [hunk 18]
```diff
@@ -0,0 +1,161 @@
...
+import * as Shim from './shim.native'
+import {debounce} from 'lodash-es'
+
+// turn on screens
+useScreens()
+
+// Options used by default on all navigators
+const defaultNavigationOptions = {
+  header: null,
+  headerLeft: hp => (
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/router-v2/router.native.js]: We'll figure out a better typing story when we deprecate the old way for sure

File: shared/router-v2/router.desktop.js [hunk 19]
```diff
@@ -0,0 +1,266 @@
...
+          <Component {...this.props} navigation={navigation} />
+        </NavigationProvider>
+      )
+    }
+
+    getNavState = () => this._navState || this.state.nav
+
+    dispatchOldAction = (old: any) => {
+      const actions = Shared.oldActionToNewActions(old, this._navigation) || []
+      actions.forEach(a => this.dispatch(a))
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/router-v2/router.desktop.js]: This pattern is a little more dangerous so I avoid it typically. It's easy to accidentally pass params you didn't expect

File: shared/router-v2/router.desktop.js [hunk 20]
```diff
@@ -0,0 +1,266 @@
...
+      }
+
+      dispatchActionEvents()
+      return false
+    }
+
+    _getScreenProps = () => this.props.screenProps
+
+    render() {
+      let navigation = this.props.navigation
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/router-v2/router.desktop.js]: Yes. This and above I wanted to keep the code very similar and not make trivial changes from the source file so it's easy to update if we need to

File: shared/wallets/routes.js [hunk 21]
```diff
@@ -114,3 +118,79 @@ const routeTree = () => {
...
+
+    return (
+      <Kb.Box2 direction="horizontal" fullHeight={true} fullWidth={true}>
+        {/* $FlowIssue */}
+        <WalletsAndDetails>
+          <SceneView
```
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply shared/wallets/routes.js]: yeah

[the remark this answers is not in the document]
Comment (buoyad) [reply shared/wallets/routes.js]: @MarcoPolo answering or do you have the same question? @chrisnojima and I talked IRL yesterday I thought the answer was no.

File: shared/wallets/routes.js (hunk 13 above)
Comment (MarcoPolo) [reply shared/wallets/routes.js]: Why not in the constructor then? I guess this is fine, just kinda weird since we don't do this like this anywhere else

File: shared/wallets/routes.js (hunk 21 above)
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply shared/wallets/routes.js]: Oh, sorry I misunderstood your question. I thought you meant that you were wondering how to display another scene here and if it was like how it was handled in the app/modal stuff like [this](https://github.com/keybase/client/pull/15577/files/8b6a888fbd50e7c5069fadfda4e7ed73589d46ec#diff-48a273a931fbc8da057c75f10207c7d9R80). I was answering yes, if you wanted to display multiple scenes (from different navigators) you could do something similar.

What you actually asked was that modals don't appear when you are here. That sounds like a bug, since we shouldn't have to do anything here to make modals appear over the app. That should all be handled in the code I linked above. Sorry for the confusion!

File: shared/router-v2/router.desktop.js [hunk 22]
```diff
@@ -0,0 +1,266 @@
...
+      }
+    }
+
+    componentDidMount() {
+      let action = this._initialAction
+      let startupState = this.state.nav
+      if (!startupState) {
+        startupState = Component.router.getStateForAction(action)
+      }
+      const dispatchActions = () =>
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/router-v2/router.desktop.js]: same comment. i'm trying to keep this the same as the base component and not making changes like this

File: shared/router-v2/router.desktop.js [hunk 23]
```diff
@@ -0,0 +1,266 @@
...
+          })
+        )
+
+      if (startupState === this.state.nav) {
+        dispatchActions()
+        return
+      }
+
+      // eslint-disable-next-line react/no-did-mount-set-state
+      this.setState({nav: startupState}, () => {
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/router-v2/router.desktop.js]: same comment. i'm trying to keep this the same as the base component and not making changes like this

File: shared/wallets/routes.js (hunk 13 above)
Comment (chrisnojima) [reply shared/wallets/routes.js]: then i'd have to store if it its in the constructor

File: shared/search/user-input/index.native.js (hunk 11 above)
Comment (chrisnojima) [reply shared/search/user-input/index.native.js]: this can likely go away, i was having issues w/ the keyboard avoiding and how it interacted w/ the animating layers. i suspected it had to do with the focusing coming / going but i think it was due to something else

File: shared/router-v2/switcheroo.js (hunk 10 above)
Comment (MarcoPolo) [reply shared/router-v2/switcheroo.js]: They suggest this? where? I think it would be good to move away from dispatching anything other than pure data (Immutable stuff and HiddenString is fine since I count it as data).

That said, this in particular seems easy to fix in the future, so I wouldn't block on this. Just wanted to point out my thoughts.

File: shared/router-v2/router.desktop.js (hunk 22 above)
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply shared/router-v2/router.desktop.js]: Yeah, sorry. I wrote these first without realizing it was copy/pasted. That reasoning makes sense 💯

File: shared/router-v2/router.shared.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply shared/router-v2/router.shared.js]: yeah, we can do the same check inside `getStateForAction`

File: shared/settings/routes.native.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply shared/settings/routes.native.js]: I forked this into `marco/DESKTOP-nav2-with-tabs-renamed` which renamed the tabs to be `tabs.X` so we can use `router.getPathAndParamsFromState` instead of our custom `getVisiblePath`

File: shared/router-v2/router.shared.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply shared/router-v2/router.shared.js]: ignoring the current name of route tree, these are our own custom routing actions. It makes more sense for them to live in `getStateForAction` (which is sort of the reducer of this navigation model) alongside the rest of the logic for handling routing actions than it does for them to live in this random spot.

[the remark this answers is not in the document]
Comment (MarcoPolo) [reply shared/router-v2/router.shared.js]: I'm talking specifically about the clearModals action here which is only used by the new nav. I think it's okay for now to keep the other things here because they will eventually go away.

File: shared/router-v2/router.shared.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (MarcoPolo) [reply shared/router-v2/router.shared.js]: yeah. What's the purpose of this check?

File: shared/chat/conversation/create-team-header/container.js (hunk 9 above)
Comment (chrisnojima) [reply shared/chat/conversation/create-team-header/container.js]: yeah it doens't work

Comment (chrisnojima) [reply shared/chat/conversation/create-team-header/container.js]: there are lots of known issues that i didn't fix. this PR is to get this in (feature flagged) so we can work on sections as a team

File: shared/router-v2/router.native.js [hunk 24]
```diff
@@ -0,0 +1,161 @@
...
+    const nav = this._nav
+    if (!nav) {
+      return
+    }
+    const path = Constants.getVisiblePath()
+
+    // We determine if we're at the root if we're at the root of our navigation hierarchy, which is slightly different if you're logged in or out
+    if (path[0].routeName === 'loggedIn') {
+      if (path[1].routeName === 'Main') {
+        if (path.length === 3) {
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/router-v2/router.native.js]: lets leave this for now and discuss a better way. this is for the back handler which likely belongs here

File: shared/constants/router2.js [hunk 25]
```diff
@@ -0,0 +1,25 @@
...
+  if (!s) return arr
+  if (!s.routes) return s
+  const route = s.routes[s.index]
+  if (!route) return arr
+  if (route.routes) return findVisibleRoute([...arr, route], route)
+  return [...arr, route]
+}
+
+// Public API
+export const getVisiblePath = (_fromNavOnly?: any) => {
```
[the remark this answers is not in the document]
Comment (chrisnojima) [reply shared/constants/router2.js]: this is a parameter thats only ever supposed to be used by the navigator itself, i'll just make a second function

**Comment by chrisnojima** [comment]
- [ ] turn ff.native off
- [ ] kick version to 320

**Comment by MarcoPolo** [comment]
Result from timings check of when the last message shows up on chat.

Run 1: 4.859
Run 2: 5.434
Run 3: 5.348
Run 4: 4.882
Run 5: 4.872

avg: 5.079

compared to an average of 4.9 on master. It's basically the same.

I also played around with various things since now we don't have to load any of the old route trees anymore from setRouteDefs, but that didn't seem to have a noticeable impact.

**Comment by MarcoPolo** [comment]
👍 🚢 🇮🇹
